### PR TITLE
Manage merge keys of top-level lists in a more consistent way

### LINKED ISSUE #102: Manage merge keys of top-level lists in a more consistent way
In the current model, all types of commands are expected to have an `id` that is mainly the key of the command in the command list.

It is defined this way:

```yaml
commands:
  - exec:
      id: "commandId"  <------ the key of the given command in the list of commands 
```

Since finally the `id` field is common to all types of command and is the key that will be used to identify a command especially while merge commands between a parent and the current devfile, it seems it might make sense to bring it one level up, like this:

```yaml
commands:
  - id: "commandId"
    exec:
      ....
```

The same stands for components: until now all types of components have a `name` field that is used as a key for identifying a component, especially during parent or plugin overriding. It is mandatory for most component types, and optional for plugins, in which case it is calculated from the definition of the plugin itself (based on id, registry url, etc ...).

Instead of: 
```yaml
components:
  - container:
      name: "tooling"
```
we would have
```yaml
components:
  - name: "tooling"
    container:
      ....
```
In this case, though, the `name` field can only be made either required or optional for all the component types, and not per-component-type. We should then probably make it mandatory for the Kubernetes CR, and optional for devfiles (but for all components). We would still loose some schema validation.
Of course, deciding that in any case we require a `name` for a component (even a plugin) would make the whole more consistent and schema-validated, but less user-friendly I admit.

So that this requires further discussion.

Related content: https://github.com/devfile/kubernetes-api/pull/98#user-content-keys

### PR DESCRIPTION
**Pull request by kadel**
### What does this PR do?

- moves component and command merge keys (`name` and `id`) to top-level
- This change requires updates to utility functions for handling unions ([union.go](https://github.com/devfile/api/blob/master/pkg/apis/workspaces/v1alpha1/union.go)). I was not able to found if this is being used somewhere. So I tried to write unit tests for it. I'm not sure if those functions actually ever worked correctly :-( 

TODO:
- [ ] update union.go and tests

### What issues does this PR fix or reference?
fixes  #102

### CODE REVIEW

**Comment by amisevsk** [comment]
@kadel I believe the union/etc. functions will be updated in https://github.com/devfile/api/pull/98, if you feel like reviewing a 10k PR ;). That PR also adds a good amount of tests.

**Comment by kadel** [comment]
> @kadel Ah yeah I can't believe I forgot that, given that I opened the blocker issue. Yes, these changes should be applied to a `v1alpha2` version -- sorry for the hassle.

No problem. I realized that after I opened this PR myself :-/ . Will update it today.

**Comment by davidfestal** [comment]
@kadel @amisevsk I merged #98 first, precisely because of the incompatibility this PR introduces in the `v1alpha1` K8S API. So the merged overriding implementation supports `v1alpha1`, and could be then updated to also support `v1alpha2` I assume.

**Comment by davidfestal** [comment]
> > @kadel Ah yeah I can't believe I forgot that, given that I opened the blocker issue. Yes, these changes should be applied to a `v1alpha2` version -- sorry for the hassle.
> 
> No problem. I realized that after I opened this PR myself :-/ . Will update it today.

So after discussing the situation again, the plan would be:
1. Create a branch (or tag) on `master` with the existing `v1alpha1` API, and let the workspace controller depend on this commit.
2. Create and merge a distinct PR that would globally rename `v1alpha1` folder and all `v1alpha1`mentions to `v1alpha2` on the `master` branch
3. Update this PR to merge it with the changes made in PR #98
4. Merge this branch on master.
5. On the workspace controller side, start working on the reintegration of the `v1alpha1` + conversion webhooks, on a distinct PR branch
5. When this last PR is merged, the workspace controller would be able to support both `v1alpha1` and `v1alpha2` with the conversion webhook.

So the main point here is that this PR should wait for the steps 1 and 2 to be finished.

**Comment by amisevsk** [comment]
> @amisevsk please complete or correct this if necessary.

The above outline is basically it. There's some grey area about how integration will work down the line, but the main idea is to "fix" `v1alpha1` in a tag/branch, and only support `v1alpha2` in the repo for now, with the goal of implementing converstion webhooks, etc. before "releasing" `v1alpha2`. The reason I think we *can* do this is because the devworkspace controller depends on the api and crds, whereas odo only depends on the APIs.

File: pkg/apis/workspaces/v1alpha1/commands.go [hunk 1]
```diff
@@ -60,8 +60,13 @@ type LabeledCommand struct {
 }
 
 // +k8s:openapi-gen=true
-// +union
 type Command struct {
+	LabeledCommand `json:",inline"`
```
Comment (davidfestal) [inline pkg/apis/workspaces/v1alpha1/commands.go:64]: I would not do it like that.
I would keep BaseCommand and LabeledCommand where they are, and would bring only the ID to the top level.

File: pkg/apis/workspaces/v1alpha1/union.go [hunk 2]
```diff
@@ -9,32 +9,43 @@ type Identified interface {
...
 func visitUnion(union interface{}, visitor interface{}) (err error) {
+
 	visitorValue := reflect.ValueOf(visitor)
 	unionValue := reflect.ValueOf(union)
 	oneMemberPresent := false
 	typeOfVisitor := visitorValue.Type()
 	for i := 0; i < visitorValue.NumField(); i++ {
 		unionMemberToRead := typeOfVisitor.Field(i).Name
 		unionMember := unionValue.FieldByName(unionMemberToRead)
-		if !unionMember.IsNil() {
```
Comment (davidfestal) [inline pkg/apis/workspaces/v1alpha1/union.go:20]: In fact the function that should be called instead of `IsNil` is `IsZero`. I already fixed that in [PR #98](https://github.com/devfile/api/pull/98/files#diff-6965f4fa88f2bd3a5d4f7ac21fb9666aR16).

File: schema-transformation-rules/devworkspace/addRequiredFields.json [hunk 3]
```diff
@@ -1,13 +1,13 @@
 [
   { 
     "op": "add",
-    "path": "/properties/spec/properties/commands/items/properties/exec/required/-",
```
Comment (davidfestal) [inline schema-transformation-rules/devworkspace/addRequiredFields.json:4]: Why is this change required ?

Comment (kadel) [reply schema-transformation-rules/devworkspace/addRequiredFields.json:4]: This patch was failing because `required` field is no longer present in `exec`

File: pkg/apis/workspaces/v1alpha1/commands.go (hunk 1 above)
Comment (kadel) [reply pkg/apis/workspaces/v1alpha1/commands.go:64]: I noticed that vscode commands don't have `label`, is that intentional?

**Comment by davidfestal** [comment]
@kadel I removed the WIP flag since I reworked this PR to:
- integrate some required refactoring of the previously-merged PR #98
- rebase on `master` (which includes being now based on the new `v1alpha2` K8S API)
- add (as the last commit of this PR) the ability to override the `starterProjects` list in a child devfile.

@amisevsk You might want to review the PR again.

File: pkg/utils/overriding/merging.go [hunk 4]
```diff
@@ -84,17 +90,21 @@ func mergeKeyedDevWorkspaceTemplateSpec(
...
+	result := workspaces.DevWorkspaceTemplateSpecContent{}
 
 	preStartCommands := sets.String{}
 	postStartCommands := sets.String{}
 	preStopCommands := sets.String{}
 	postStopCommands := sets.String{}
 
 	for _, content := range allContents {
+
+		// TODO: Generate a function to implement this merging logic directly from
```
Comment (sleshchenko) [inline pkg/utils/overriding/merging.go:102]: Such TODOs in code are not well visible. It think it makes sense to register an issue for that instead of in addition.

File: pkg/utils/overriding/merging.go [hunk 5]
```diff
@@ -40,7 +41,10 @@ func ensureNoConflictsWithPlugins(mainContent *keyedDevWorkspaceTemplateSpecCont
 		return "unknown"
 	}
 
-	allSpecs := append([]*keyedDevWorkspaceTemplateSpecContent{mainContent}, pluginFlattenedContents...)
+	allSpecs := []workspaces.TopLevelListContainer{mainContent}
+	for _, plugipluginFlattenedContent := range pluginFlattenedContents {
```
Comment (amisevsk) [inline pkg/utils/overriding/merging.go:45]:
```suggestion
	for _, pluginFlattenedContent := range pluginFlattenedContents {
```
:)

File: pkg/utils/overriding/merging.go [hunk 6]
```diff
@@ -61,20 +65,23 @@ func ensureNoConflictsWithPlugins(mainContent *keyedDevWorkspaceTemplateSpecCont
...
+// MergeDevWorkspaceTemplateSpec implements the merging logic of a main devfile content with flattened, already-overriden parent devfiles or plugins.
+// On an `main` `DevfileWorkspaceTemplateSpec` (which is the core part of a devfile, without the `apiVersion` and `metadata`),
+// it allows adding all the new overriden elements provided by flattened parent and plugins
+//
+// It is not allowed for to have duplicate (== with same key) commands, components or projects between the main content and the parent or plugins.
+// An error would be thrown
+//
+// The result is a transformed `DevfileWorkspaceTemplateSpec` object, that does not contain any `plugin` component
+// (since they are expected to be provided as flattened overriden devfiles in teh arguments)
+func MergeDevWorkspaceTemplateSpec(
```
Comment (amisevsk) [inline pkg/utils/overriding/merging.go:77]: Since it's exported, this function should go at the top of the file.

File: pkg/utils/overriding/merging.go [hunk 7]
```diff
@@ -61,20 +65,23 @@ func ensureNoConflictsWithPlugins(mainContent *keyedDevWorkspaceTemplateSpecCont
...
-	pluginFlattenedContents ...*keyedDevWorkspaceTemplateSpecContent) (*keyedDevWorkspaceTemplateSpecContent, error) {
+// MergeDevWorkspaceTemplateSpec implements the merging logic of a main devfile content with flattened, already-overriden parent devfiles or plugins.
+// On an `main` `DevfileWorkspaceTemplateSpec` (which is the core part of a devfile, without the `apiVersion` and `metadata`),
+// it allows adding all the new overriden elements provided by flattened parent and plugins
+//
+// It is not allowed for to have duplicate (== with same key) commands, components or projects between the main content and the parent or plugins.
+// An error would be thrown
+//
+// The result is a transformed `DevfileWorkspaceTemplateSpec` object, that does not contain any `plugin` component
+// (since they are expected to be provided as flattened overriden devfiles in teh arguments)
```
Comment (amisevsk) [inline pkg/utils/overriding/merging.go:76]:
```suggestion
// MergeDevWorkspaceTemplateSpec implements the merging logic of a main devfile content with flattened, already-overridden parent devfiles or plugins.
// On a `main` `DevWorkspaceTemplateSpec` (which is the core part of a devfile, without the `apiVersion` and `metadata`),
// it allows adding all the new overridden elements provided by flattened parent and plugins
//
// Returns non-nil error if there are duplicate (== with same key) commands, components or projects between the 
// main content and the parent or plugins. 
//
// The result is a transformed `DevWorkspaceTemplateSpec` object, that does not contain any `plugin` component
// (since they are expected to be provided as flattened overridden devfiles in the arguments)
```

File: pkg/utils/overriding/merging.go [hunk 8]
```diff
@@ -84,45 +91,49 @@ func mergeKeyedDevWorkspaceTemplateSpec(
...
-		if content.Commands != nil && len(content.Commands) > 0 {
-			if result.Commands == nil {
-				result.Commands = []keyedCommand{}
-			}
-			for _, command := range content.Commands {
-				result.Commands = append(result.Commands, command)
-			}
+	topLevelListsNames := result.GetToplevelLists()
+	resultValue := reflect.ValueOf(&result).Elem()
+	componentType := reflect.TypeOf(workspaces.Component{})
```
Comment (amisevsk) [inline pkg/utils/overriding/merging.go:103]: There has to be a better way to do this than by reflection, the loops below are *so* complicated. I think it'd even be simpler to have copy-pasted code for each case here.

File: pkg/utils/overriding/keys.go [hunk 9]
```diff
@@ -6,138 +6,33 @@ import (
...
-	}
-
-	if keyedSpec.Commands != nil && len(keyedSpec.Commands) > 0 {
-		result.Commands = []workspaces.Command{}
-	}
-
-	if keyedSpec.Components != nil && len(keyedSpec.Components) > 0 {
-		result.Components = []workspaces.Component{}
-	}
+type checkFn func(elementType string, keysSets []sets.String) []error
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/utils/overriding/keys.go:9]: There is an implicit assumption here that the `keysSets` passed to `checkFn` will maintain some sort of order from the `...workspaces.TopLevelListContainer` passed in to `checkKeys`

File: pkg/apis/workspaces/v1alpha1/commands.go (hunk 1 above)
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha1/commands.go]: yes, because the label is expected to be in the VSCode raw json definition afair.

File: pkg/utils/overriding/test-fixtures/merges/events-merge/main.yaml [hunk 10]
```diff
@@ -2,8 +2,8 @@ parent:
   uri: "anyParent"
 components:
   - plugin:
-      name: "theOnlyPlugin"
       uri: "aCustomLocation"
+    name: "theOnlyPlugin"
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/utils/overriding/test-fixtures/merges/events-merge/main.yaml:6]: I thought about that, but the generated expected results will follow the typical YAML serialization field order, which is the alphabetical one. So to ease the comparison between input yamls and expected output yaml,I kept the alphabetical order in all the examples.

File: pkg/apis/workspaces/v1alpha2/keyed.go [hunk 11]
```diff
@@ -1,108 +1,11 @@
...
-			key = command.Id
-			return nil
-		},
-		VscodeTask: func(command *VscodeConfigurationCommand) error {
-			key = command.Id
-			return nil
-		},
-	})
-	return key, err
+type TopLevelListContainer interface {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/keyed.go:9]: The structs that satisfy this interface can be found in [the `keyed-definitions` source file](https://github.com/kadel/kubernetes-api/blob/top-level-merge-key/pkg/apis/workspaces/v1alpha2/keyed_definitions.go#L23) which is expected to be automatically generated from the main API structs in the future.

File: pkg/utils/overriding/test-fixtures/merges/duplicate-with-plugin/plugin.yaml [hunk 12]
```diff
@@ -1,4 +1,4 @@
 components:
   - container:
       image: "aValue"
-      name: "existingInPlugin"
+    name: "existingInPlugin"
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/utils/overriding/test-fixtures/merges/duplicate-with-plugin/plugin.yaml]: fixed

File: pkg/utils/overriding/test-fixtures/patches/add-command-and-component/patch.yaml [hunk 13]
```diff
@@ -1,21 +1,21 @@
...
 
   - exec:
-      id: "commandToAdd"
       commandLine: "commandToAddCommandLine"
+    id: "commandToAdd"
 components:
-  - volume:
-      name: newVolume
+  - volume: {}
+    name: newVolume
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/utils/overriding/test-fixtures/patches/add-command-and-component/patch.yaml]: Fixed

File: pkg/apis/workspaces/v1alpha2/keyed.go [hunk 14]
```diff
@@ -1,108 +1,11 @@
...
-			return nil
-		},
-		Volume: func(vol *VolumeComponent) error {
-			key = vol.Name
-			return nil
-		},
-	})
-	return key, err
-}
+type TopLevelLists map[string][]string
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/keyed.go]: In this commit https://github.com/devfile/api/pull/122/commits/3c18672f7de1f419ae53b732b16804c0b38b284c, I added documentation and also use a new `KeyedList` type instead of `[]string` as the type of map items.

[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/keyed.go]: Do those docs make things clearer ?

File: pkg/apis/workspaces/v1alpha2/keyed_definitions.go [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+	}
+}
+
+func (container ParentOverrides) GetToplevelLists() TopLevelLists {
+	return TopLevelLists{
+		"Projects":        extractKeys(container.Projects),
+		"StarterProjects": extractKeys(container.StarterProjects),
+		"Components":      extractKeys(container.Components),
+		"Commands":        extractKeys(container.Commands),
+	}
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/keyed_definitions.go:45]: The question should be answered by the documentation added in the following commit: https://github.com/devfile/api/pull/122/commits/3c18672f7de1f419ae53b732b16804c0b38b284c#diff-1a6c626c12f4a1882d43befff929fd51R35

[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/keyed_definitions.go:45]: But keep in mind that this file is expected to be generated automatically during the build process from the main K8S API objects and devfile-specific annotations.

File: pkg/utils/overriding/merging.go (hunk 6 above)
Comment (davidfestal) [reply pkg/utils/overriding/merging.go:23]: Fixed

File: pkg/utils/overriding/keys.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/utils/overriding/keys.go:9]: Is it clearer with the documentation provided in commithttps://github.com/devfile/api/pull/122/commits/f5bcbc623c4b0631f46e1bacea35ac477cadf610 ?

File: pkg/utils/overriding/test-fixtures/merges/events-merge/main.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/utils/overriding/test-fixtures/merges/events-merge/main.yaml:6]: :+1: makes sense

File: pkg/apis/workspaces/v1alpha2/keyed.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/apis/workspaces/v1alpha2/keyed.go:39]: :+1:

File: pkg/apis/workspaces/v1alpha2/keyed.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/apis/workspaces/v1alpha2/keyed.go]: Yeah that helps, thanks.

File: pkg/apis/workspaces/v1alpha2/keyed_definitions.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/apis/workspaces/v1alpha2/keyed_definitions.go:45]: Ok, makes sense.

File: pkg/utils/overriding/keys.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/utils/overriding/keys.go:9]: It makes more sense, but I still think the structure of `checkFn` and `checkKeys` is strange -- `checkKeys` is doing too much. It's effectively

1. Merging a `[]map[string][]string` into `map[string][][]string` based on map keys
2. Applying `checkFn` to each `(string, []string)` pair in the flattened map to map the `[]string` to error

While this is a very general function, it's also not a "standard" function one would expect to see, so understanding how it works requires additional reading. While more verbose, I think splitting it into two functions would make it easier ot understand:

```go
func mergeTopLevelLists(...map[string][]string) map[string][]sets.String {...}

func checkTopLevelLists(checkFn, map[string][]sets.String) error {...}
```

Restricting it further based on current uses (so maybe not valid if I'm missing a case), I would go so far as to remove `checkTopLevelLists` and `checkFn` altogether, since the `elementType` argument for `checkFn` is only used for error messages:

```go
func ensureNoConflictWithParent(...) []error {
  mergedTopLevelLists := mergetopLevelLists(mainContent, parentflattenedContent)
  for elementType, listElem := range mergedTopLevelLists {
    mainKeys := listElem[0]
    parentOrPluginKeys := listElem[1]
    overriddenElementsInMainContent := mainKeys.Intersection(parentOrPluginKeys)
    if overriddenElementsInMainContent.Len() > 0 {
      return []error{fmt.Errorf("Some %s are already defined in parent: %s. "+
        "If you want to override them, you should do it in the parent scope.",
        elementType,
        strings.Join(overriddenElementsInMainContent.List(), ", "))}
    }
    return []error{}
  }
}
```
it's a bit more repetition (since we have to write the loop over `(string, [][]string`) pairs, but i would argue easier to understand. We could go further in this case, and iterate over key-value pairs in `mainContent` directly, to make our check more explicit.

This also raises a question about the current implementation: if we're to assume that `keysSets[0]` is potentially special (e.g. it is the `mainContent` keys above), how do we detect the difference between the case where a subsequent element gets merged down and introduces a `listType` not present in `mainContent`? For example, with `ensureNoConflictsWithPlugins`, if `mainContent` doesn't contain any `Components`, won't it seem like the first `pluginFlattenedContent`'s components represent components in `mainContent`? Perhaps more clearly, the code can't distinguish between
```
topLevelListContainer1 {
  "Commands": "hello"
  "Components": "I'm a component"
}
topLevelListContainer2 {
  "Commands": "hello2"
}
```
and 
```
topLevelListContainer1 {
  "Commands": "hello"
}
topLevelListContainer2 {
  "Commands": "hello2"
  "Components": "I'm a component"
}
```
since they both merge to
```
map[string][]sets.String {
  "Commands": [{"hello"}, {"hello2"}],
  "Components": [{"I'm a component"}],
}
```

File: pkg/utils/overriding/merging.go (hunk 8 above)
Comment (davidfestal) [reply pkg/utils/overriding/merging.go]: I did an attempt at simplifying it in commit: https://github.com/devfile/api/pull/122/commits/88d00ae9e2509417ac4f0c9118b5f2c52488e00b

File: pkg/utils/overriding/keys.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/utils/overriding/keys.go:9]: I suppose the last concern is a non-issue, because of they way `keyed_definitions.go` is written, but we're still relying on an implicit implementation in another package here.

**Comment by amisevsk** [comment]
@davidfestal I've added https://github.com/devfile/api/pull/122/commits/ef7ba598c18373a37db0a5d42ccc0412a5890f66 to simplify `checkKeys` as we discussed (GitHub has once again chosen to break the readability of the diff by matching close parens :upside_down_face:). PTAL if you have a moment.

cc: @kadel

File: pkg/utils/overriding/merging.go (hunk 8 above)
Comment (amisevsk) [reply pkg/utils/overriding/merging.go]: LGTM, thanks.

File: pkg/apis/workspaces/v1alpha1/commands.go (hunk 1 above)
Comment (kadel) [reply pkg/apis/workspaces/v1alpha1/commands.go]: ok makes sense

**Review by kadel — COMMENTED** [review]
LGTM 👍
